src/gar.lua

```diff
@@ -1191,4 +1196,259 @@ function gar.delegateAllowedToStake(delegateAddress, gateway)
+			utils.filterDictionary(gar.getGatewaysUnsafe(), function(_, gateway)
+				return gateway.delegates[fundingPlan.address] ~= nil
+			end),
+			-- extract only the essential gateway fields, copying tables so we don't mutate references
+			function(acc, gatewayAddress, gateway)
+				local totalEpochsGatewayPassed = gateway.stats.passedEpochCount or 0
+				local totalEpochsParticipatedIn = gateway.stats.totalEpochCount or 0
+				local gatewayRewardRatioWeight = (1 + totalEpochsGatewayPassed) / (1 + totalEpochsParticipatedIn)
+				local delegate = utils.deepCopy(gateway.delegates[fundingPlan.address])
+				delegate.excessStake = math.max(0, delegate.delegatedStake - gateway.settings.minDelegatedStake)
+				delegate.gatewayAddress = gatewayAddress
+				table.insert(acc, {
+					totalDelegatedStake = gateway.totalDelegatedStake, -- for comparing gw total stake
+					gatewayRewardRatioWeight = gatewayRewardRatioWeight, -- for comparing gw performance
+					delegate = delegate,
+					startTimestamp = gateway.startTimestamp, -- for comparing gw tenure
+				})
+				return acc
+			end,
+			{}
+		),
+		{
+			{
+				order = "desc",
+				field = "delegate.excessStake",
+			},
+			{
+				order = "asc",
+				field = "gatewayRewardRatioWeight",
+			},
+			{
+				order = "desc",
+				field = "totalDelegatedStake",
+			},
+			{
+				order = "desc",
+				field = "startTimestamp",
+			},
+		}
+	)
+
+	-- simulate drawing down excess stakes until the remaining balance is satisfied OR excess stakes are exhausted
+	local gatewayIndex, nextGateway = next(gatewaysInfo)
+	while fundingPlan.shortfall > 0 and nextGateway do
+		local excessStake = nextGateway.delegate.excessStake
+		local stakeToDraw = math.min(excessStake, fundingPlan.shortfall)
+		fundingPlan["stakes"][nextGateway.delegate.gatewayAddress] = {
+			delegatedStake = stakeToDraw,
+			vaults = {}, -- set up vault spend tracking now while we're passing through
+		}
+		fundingPlan.shortfall = fundingPlan.shortfall - stakeToDraw
+		nextGateway.delegate.delegatedStake = nextGateway.delegate.delegatedStake - stakeToDraw
+		-- maintain consistency
+		nextGateway.delegate.excessStake = excessStake - stakeToDraw
+		nextGateway.totalDelegatedStake = nextGateway.totalDelegatedStake - stakeToDraw
+		gatewayIndex, nextGateway = next(gatewaysInfo, gatewayIndex)
+	end
+	return gatewaysInfo
+end
+
+function planVaultsDrawdown(fundingPlan, gatewaysInfo)
+	-- simulate drawing down vaults until the remaining balance is satisfied OR vaults are exhausted
+	local vaults = utils.sortTableByFields(
+		utils.reduce(gatewaysInfo, function(acc, _, gatewayInfo)
+			for vaultId, vault in pairs(gatewayInfo.delegate.vaults) do
+				table.insert(acc, {
+					vaultId = vaultId,
+					gatewayAddress = gatewayInfo.delegate.gatewayAddress,
+					endTimestamp = vault.endTimestamp,
+					balance = vault.balance,
+				})
+			end
+			return acc
+		end, {}),
+		{
+			{
+				order = "asc",
+				field = "endTimestamp",
+			},
+		}
+	)
```

Comment:
the nested sort/reduce are hard to follow. may be helpful to breakup, at the sacrifice of memory footprint

Reply:
b7f254d